### PR TITLE
Server side implementation of Question Detail Page of Test Conduct

### PR DESCRIPTION
**Pull request by Deepankar-promact**
**Fixed Issues**

Added Models, APIs and their respective unit test.

**Files Changed**

TestConductController.cs - Added controller function to add and get answers from the database.
TestController.cs - Added controller function to get questions in a test by test id.
ITestRepository.cs - Added definition of repository function.
TestRepository.cs - Added Implementation of GetTestQuestionByTestIdAsync function.
ITestConductRepository.cs - Added definition of repository function.
TestConductRepository.cs - Added Implementation of add and get answer function.
TestConductRepositoryTest.cs - Added Unit Testing.
TestRepositoryTest.cs - Added Unit Testing.
Answers.cs - Model to store answers.
AttendeeAnswers.cs - Model to store answer as a key-value pair.
TestConduct.cs - Modle to store the question attempted and their status.

**Checks**

- [x] Naming Conventions
- [x] Server side code comments (proper English, grammar and no spelling mistakes)
- [x] Client side code comments (proper English, grammar and no spelling mistakes)
- [x] Unused variables, methods, blank spaces and name spaces. 
- [x] Optimal Code and code formatting
- [x] Commit History is proper, no merge is done. 
- [x] Commit/pull request messages should be proper to justify your feature
- [x] All test cases are executed provided by tester.

**Known Issues**

- No issue

**Comments**

NA

---
This change is [Reviewable](https://reviewable.io/reviews/promact/trappist/337)

### CODE REVIEW

File: Trappist/src/Promact.Trappist.Core/Controllers/TestConductController.cs [hunk 1]
```diff
@@ -1,24 +1,33 @@
...
 using Promact.Trappist.DomainModel.Models.TestConduct;
 using Promact.Trappist.Repository.TestConduct;
+using Promact.Trappist.Repository.Tests;
+using Promact.Trappist.Web.Models;
 using System.Threading.Tasks;
 
 namespace Promact.Trappist.Core.Controllers
 {
+    [Authorize]
     [Route("api/conduct")]
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Core/Controllers/TestConductController.cs]: route name should be api/testconduct

File: Trappist/src/Promact.Trappist.Core/Controllers/TestConductController.cs [hunk 2]
```diff
@@ -1,24 +1,33 @@
...
     public class TestConductController : Controller
     {
         #region Private Variables
         #region Dependencies
         private readonly ITestConductRepository _testConductRepository;
+        private readonly ITestsRepository _testRepository;
+        private readonly UserManager<ApplicationUser> _userManager;
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.Core/Controllers/TestConductController.cs]: MAJOR Remove the '_userManager' field and declare it as a local variable in the relevant methods. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS1450)

**Comment by chintans** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: Trappist/src/Promact.Trappist.DomainModel/DbContext/TrappistDbContext.cs [hunk 3]
```diff
@@ -70,7 +70,9 @@ protected override void OnConfiguring(DbContextOptionsBuilder optionBuilder)
         public DbSet<TestAttendees> TestAttendees { get; set; }
         public DbSet<TestCategory> TestCategory { get; set; }
         public DbSet<TestQuestion> TestQuestion { get; set; }
-
+        public DbSet<TestConduct> TestConduct { get; set; }
+        public DbSet<Answers> Answers { get; set; }
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.DomainModel/DbContext/TrappistDbContext.cs]: Answers are generalized. Can we specify more details?

File: Trappist/src/Promact.Trappist.DomainModel/Models/TestConduct/Answers.cs [hunk 4]
```diff
@@ -0,0 +1,16 @@
+﻿using System.ComponentModel.DataAnnotations.Schema;
+
+namespace Promact.Trappist.DomainModel.Models.TestConduct
+{
+    public class Answers : BaseModel
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.DomainModel/Models/TestConduct/Answers.cs]: same as above

File: Trappist/src/Promact.Trappist.Test/TestConduct/TestConductRepositoryTest.cs [hunk 5]
```diff
@@ -164,6 +165,47 @@ public TestConductRepositoryTest(Bootstrap bootstrap) : base(bootstrap)
...
         /// <summary>
+        /// Test to add answer
+        /// </summary>
+        [Fact]
+        public async Task AddAnswersAsyncTest()
+        {
+            var testAttendee = InitializeTestAttendeeParameters();
+            await CreateTestAsync();
+            await _testConductRepository.RegisterTestAttendeesAsync(testAttendee, _stringConstants.MagicString);
+            var attendeeId = await _trappistDbContext.TestAttendees.Where(x=>x.Email==testAttendee.Email).Select(x=>x.Id).FirstOrDefaultAsync();
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Test/TestConduct/TestConductRepositoryTest.cs]: formatting is missing on this line.

File: Trappist/src/Promact.Trappist.Web/Migrations/20170501095838_AddedModelForConductingTest.cs [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+using Microsoft.EntityFrameworkCore.Migrations;
+using Microsoft.EntityFrameworkCore.Metadata;
+
+namespace Promact.Trappist.Web.Migrations
+{
+    public partial class AddedModelForConductingTest : Migration
+    {
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Web/Migrations/20170501095838_AddedModelForConductingTest.cs]: extra spaces exist

File: Trappist/src/Promact.Trappist.Web/wwwroot/app/conduct/conduct.service.ts [hunk 7]
```diff
@@ -3,7 +3,7 @@ import { HttpService } from '../core/http.service';
 
 @Injectable()
 export class ConductService {
-    private registerTestAttendeeApiUrl: string = '/api/conduct/';
+    private registerTestAttendeeApiUrl: string = '/api/testconduct/';
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/conduct/conduct.service.ts]: make it conduct. api route is wrong

Comment (Deepankar-promact) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/conduct/conduct.service.ts]: I had changed the API route to `api/testconduct` as per your previous review. Please verify the route.

Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/conduct/conduct.service.ts]: please revert the route. No changes required

File: Trappist/src/Promact.Trappist.Core/Controllers/TestConductController.cs [hunk 8]
```diff
@@ -40,6 +45,38 @@ public TestConductController(ITestConductRepository testConductRepository)
             return BadRequest();
         }
 
+        /// <summary>
+        /// Adds answer to the Database as a Key-Value pair
+        /// </summary>
+        /// <param name="attendeeId">Id of Test Attendee</param>
+        /// <param name="answer">Answer submitted</param>
+        [HttpPut("answer/{attendeeId}")]
+        public async Task<IActionResult> AddAnswerAsync([FromRoute]int attendeeId, [FromBody]object answer)
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Core/Controllers/TestConductController.cs]: Please create some class. Do not use object directly.

File: Trappist/src/Promact.Trappist.Repository/Test/TestsRepository.cs [hunk 9]
```diff
@@ -257,6 +257,28 @@ public TestsRepository(TrappistDbContext dbContext, IGlobalUtil util, IStringCon
...
                 return null;
         }
+
+        public async Task<ICollection<QuestionAC>> GetTestQuestionByTestIdAsync(int testId)
+        {
+            var question = new QuestionAC();
+            var questionList = new List<QuestionAC>();
+
+            await _dbContext.TestQuestion
+                .Include(x=>x.Question)
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/Test/TestsRepository.cs]: please use visual studio formatting here. no line is formatted properly.

File: Trappist/src/Promact.Trappist.Repository/TestConduct/ITestConductRepository.cs [hunk 10]
```diff
@@ -35,5 +35,18 @@ public interface ITestConductRepository
         /// <param name="testLink">link to conduct a particular test</param>
         /// <returns>instructions for a particular test</returns>
         Task<TestInstructionsAC> GetTestInstructionsAsync(string testLink);
+        
+        /// <summary>
+        /// Stores answers as Key-Value pair in the Database
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/TestConduct/ITestConductRepository.cs]: store answers as Key-Value pair in the Database.
we use the singular in a case of a bunch of objects.

File: Trappist/src/Promact.Trappist.Repository/TestConduct/ITestConductRepository.cs [hunk 11]
```diff
@@ -35,5 +35,18 @@ public interface ITestConductRepository
...
+        /// </summary>
+        /// <param name="attendeeId">Id of Test Attendee</param>
+        Task AddAnswerAsync(int attendeeId, string answers);
+
+        /// <summary>
+        /// Gets answers from the Database
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/TestConduct/ITestConductRepository.cs]: Get answers from the Database.

File: Trappist/src/Promact.Trappist.Repository/TestConduct/TestConductRepository.cs [hunk 12]
```diff
@@ -72,6 +72,29 @@ public TestConductRepository(TrappistDbContext dbContext)
             var isTestLinkExist = await (_dbContext.Test.AnyAsync(x => (x.Link == magicString)));
             return isTestLinkExist;
         }
+
+        public async Task AddAnswerAsync(int attendeeId, string answers)
+        {
+            if(await _dbContext.AttendeeAnswers.AnyAsync(x=>x.Id == attendeeId))
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/TestConduct/TestConductRepository.cs]: same code is not formatted. Please use ctrl K + D in all files.

File: Trappist/src/Promact.Trappist.Test/TestConduct/TestConductRepositoryTest.cs [hunk 13]
```diff
@@ -164,6 +165,47 @@ public TestConductRepositoryTest(Bootstrap bootstrap) : base(bootstrap)
...
         /// <summary>
+        /// Test to add answer
+        /// </summary>
+        [Fact]
+        public async Task AddAnswersAsyncTest()
+        {
+            var testAttendee = InitializeTestAttendeeParameters();
+            await CreateTestAsync();
+            await _testConductRepository.RegisterTestAttendeesAsync(testAttendee, _stringConstants.MagicString);
+            var attendeeId = await _trappistDbContext.TestAttendees.Where(x => x.Email == testAttendee.Email).Select(x => x.Id).FirstOrDefaultAsync();
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Test/TestConduct/TestConductRepositoryTest.cs]: string comparison should be done with equals method.

File: Trappist/src/Promact.Trappist.Test/TestConduct/TestConductRepositoryTest.cs [hunk 14]
```diff
@@ -164,6 +165,47 @@ public TestConductRepositoryTest(Bootstrap bootstrap) : base(bootstrap)
...
+        /// <summary>
+        /// Test to get answer
+        /// </summary>
+        [Fact]
+        public async Task GetAnswersAsyncTest()
+        {
+            var testAttendee = InitializeTestAttendeeParameters();
+            await CreateTestAsync();
+            await _testConductRepository.RegisterTestAttendeesAsync(testAttendee, _stringConstants.MagicString);
+            var attendeeId = await _trappistDbContext.TestAttendees.Where(x => x.Email == testAttendee.Email).Select(x => x.Id).FirstOrDefaultAsync();
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Test/TestConduct/TestConductRepositoryTest.cs]: same as above.

File: Trappist/src/Promact.Trappist.DomainModel/ApplicationClasses/TestConduct/TestAnswerAC.cs [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+using System.ComponentModel.DataAnnotations;
+
+namespace Promact.Trappist.DomainModel.ApplicationClasses.TestConduct
+{
+    public class TestAnswerAC
+    {
+        [Required]
+        public int QuestionId { get; set; }
+
+        public int[] OptionChoice { get; set; }
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.DomainModel/ApplicationClasses/TestConduct/TestAnswerAC.cs]: IEnumerable will be good than array.

File: Trappist/src/Promact.Trappist.Repository/Test/TestsRepository.cs [hunk 16]
```diff
@@ -257,6 +257,29 @@ public TestsRepository(TrappistDbContext dbContext, IGlobalUtil util, IStringCon
...
+                .Include(x => x.Question)
+                .ThenInclude(x => x.SingleMultipleAnswerQuestion)
+                .ThenInclude(x => x.SingleMultipleAnswerQuestionOption)
+                .Include(x => x.Question.CodeSnippetQuestion)
+                .ForEachAsync(x =>
+                {
+                    question.Question = Mapper.Map<QuestionDetailAC>(x.Question);
+                    question.CodeSnippetQuestion = Mapper.Map<CodeSnippetQuestionAC>(x.Question.CodeSnippetQuestion);
+                    question.SingleMultipleAnswerQuestion = Mapper.Map<SingleMultipleAnswerQuestionAC>(x.Question.SingleMultipleAnswerQuestion);
+                    //Hinding answers
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/Test/TestsRepository.cs]: What is the meaning of Hinding?

**Pull request merged**